notebooks/nearest_neighbors.livemd

```diff
@@ -0,0 +1,255 @@
+Figure 3: A process of creating a tree in Random Projection Forest. Source: http://www.math.umassd.edu/~dyan/rpKNN.png?attredirects=0
+
+<!-- livebook:{"break_markdown":true} -->
+
+Here's a brief explanation of LargeVis. First, we compute an approximated nearest neighbors graph using a Random Projection Forest. Then, we refine the graph by examining the neighbors of each point's neighbors for a fixed number of iterations. Now, let's use LargeVis in practice.
+
+In this example, we will use a dataset on online news popularity. This dataset summarizes a diverse set of features about articles published by Mashable over a two-year period. To read more about this dataset, click [here](https://archive.ics.uci.edu/dataset/332/online+news+popularity). Usually, the dataset is used to predict the number of shares on social networks, but we'll use it differently. KNN and Approximated Nearest Neighbors can be valuable in data mining, particularly in business scenarios. Imagine you're an editor at an online news agency, and you want to create a series of articles that boost your reach. One strategy could be to examine groups of the most popular articles and analyze them to identify specific patterns. We can use Approximated Nearest Neighbors to do this. For each news article, we'll calculate its nearest neighbors and then sort these groups by the total number of shares they receive on social networks.
+
+The first step is to load the dataset from the Internet archive.
+
+```elixir
+digits_url =
+  "https://archive.ics.uci.edu/static/public/332/online+news+popularity.zip"
+
+raw_data =
+  Req.get!(digits_url).body
+```
+
+Next, we load the data into a dataframe.
+
+```elixir
+{:ok, [_, {_, data}]} = :zip.extract(raw_data, [:memory])
+data = String.replace(data, ", ", ",")
+df_data = DF.load_csv!(data, delimiter: ",")
+```
+
+Next, we split the data into two parts: the target variable (shares, or $y$) and the rest of the data ($x$). We separate the shares from $x$ to avoid introducing bias. We also remove URLs because tensors do not support non-numerical data types, and we exclude the timedelta feature, which is an auxiliary feature used in creating the dataset. Finally, we apply standard scaling to $x$.
+
+```elixir
+x = df_data |> DF.discard(["url", "timedelta", "shares"]) |> Nx.stack(axis: 1)
+y = Nx.stack(df_data[["shares"]], axis: 1)
+
+x = Scholar.Preprocessing.StandardScaler.fit_transform(x)
+{x, y}
+```
+
+Now, let's move on to computing the approximated nearest neighbors. Notice that this dataset has significantly more features than the previous one (58 compared to 7).
+
+```elixir
+num_neighbors = 6
+{indices, _dist} = Scholar.Neighbors.LargeVis.fit(x, num_neighbors: num_neighbors)
+```
+
+Finally, we compute the nearest neighbors and sort them in descending order based on the sum of shares.
+
+```elixir
+shares_of_neighbors = Nx.take(y, indices) |> Nx.squeeze(axes: [-1]) |> Nx.sum(axes: [-1])
+ord = Nx.argsort(shares_of_neighbors, direction: :desc)
```

Comment:
Is sort needed here? I think they are already sorted in descending order with respect to the distance:
https://github.com/elixir-nx/scholar/blob/main/lib/scholar/neighbors/utils.ex#L38

Reply:
Sort here is by the sum of shares

Reply:
Oh I see, sorry. :)